eglfs: Allow scaling down the window surface

giucam opened the pull request:
This patch introduces the QT_QPA_EGLFS_SCALE environment variable.
When set to an integer number > 1 it will scale the draws window
content down by that scale factor.

adenexter commented on src/plugins/platforms/eglfs/qeglfscontext.cpp:
```diff
@@ -48,15 +48,212 @@
+
+QEglFSContext::~QEglFSContext()
 {
+    delete m_blitter;
```
delete m_fbo as well?

Or possibly use QScopedPointer for m_blitter and m_fbo.

adenexter commented:
The scaling works correctly for me, but I'm seeing a couple of scene graph items that aren't rendering when a scale is specified.

Specifically the glass effect which may be because the fragment shader samples screen coordinates.  And the launcher background when it is partially open, which is a little harder to explain.

sletta commented on src/plugins/platforms/eglfs/qeglfscontext.cpp:
```diff
@@ -48,15 +48,212 @@
+        initializeOpenGLFunctions();
+        m_blitProgram = new QOpenGLShaderProgram();
+        m_blitProgram->addShaderFromSourceCode(QOpenGLShader::Vertex, "attribute vec4 position;\n\
+                                                                    attribute vec4 texCoords;\n\
```
vec2

sletta commented on src/plugins/platforms/eglfs/qeglfscontext.cpp:
```diff
@@ -48,15 +48,212 @@
+                                                                    void main()\n\
+                                                                    {\n\
+                                                                        gl_Position = position;\n\
+                                                                        outTexCoords = texCoords.xy;\n\
```
if texCoords is a vec2, then you can use only texCoords, without .xy

sletta commented on src/plugins/platforms/eglfs/qeglfscontext.cpp:
```diff
@@ -48,15 +48,212 @@
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, m_minFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, m_magFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, m_wrapS);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, m_wrapT);
```
You didn't strictly touch the wrap modes

giucam replied:
I don't see what you mean...

sletta replied:
In the blit function, you are only setting MIN and MAG filter. You donæt touch the wrap modes, so you don't need to restore them either.

giucam replied:
Ah, right.

giucam commented:
I've noticed the rendering problems, but i thought they probably are mesa bugs... iirc the emulator uses mesa 9.2 which is pretty old.
I wouldn't know how to explain them otherwise.

adenexter commented:
It could be mesa, I certainly don't have a better explanation.

sletta commented:
LGTM

giucam committed:
MER: eglfs: Allow scaling down the window surface

This patch introduces the QT_QPA_EGLFS_SCALE environment variable.
When set to an integer number > 1 it will scale the draws window
content down by that scale factor.

This is only needed for the Sailfish SDK emulator, for now and i'm
not sure this belongs upstream.
